daemon/containerd/image_list.go

```diff
@@ -355,6 +358,11 @@ func (i *ImageService) imageSummary(ctx context.Context, img images.Image, platf
 
 		i.containers.ApplyAll(func(c *container.Container) {
 			if c.ImageManifest != nil && c.ImageManifest.Digest == target.Digest {
+				// ApplyAll calls the function asynchronously for each
+				// container in a separate goroutine, so we need to sync
+				// mutations to the summary.
+				mut.Lock()
+				defer mut.Unlock()
 				mfstSummary.ImageData.Containers = append(mfstSummary.ImageData.Containers, c.ID)
```

Comment:
slightly wondering; now that we have the locking.. are we effectively defeating the whole purpose of the `ApplyAll()` logic to use a goroutine? Given that the code in the goroutine is .. basically “nothing” (check the digest and append), Would it make sense to instead just use a regular loop? just loop over `container.List()`; https://github.com/moby/moby/blob/d8f079deb6d9faf6540b756baec42ede172cd53e/container/memory_store.go#L44-L50

Only downside from that one is that it also sorts the containers (which we don’t need I guess)